src/mongocrypt-endpoint-private.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MONGOCRYPT_ENDPOINT_PRIVATE_H
+#define MONGOCRYPT_ENDPOINT_PRIVATE_H
+
+#include <bson/bson.h>
+
+#include "mongocrypt-status-private.h"
+
+typedef struct {
+   /* e.g. https://kevin.keyvault.azure.net:443/path/path/?query=value */
+   char *original;
+   char *protocol;  /* e.g. https */
+   char *host;      /* e.g. kevin.keyvault.azure.net */
+   char *port;      /* e.g. 443 */
+   char *domain;    /* e.g. keyvault.azure.net */
+   char *subdomain; /* e.g. kevin */
+   char *path;      /* e.g. path/path */
+   char *query;     /* e.g. query=value */
+   /* normalized is the form that should be returned to drivers. */
+   char *normalized; /* e.g. kevin.keyvault.azure.net:443 */
```

Comment:
`normalized` seems to be a concatenation of `host` and `port`. Will `host` and `port` ever change? If so, this might not be updated to reflect that change and we should probably get this info from a function that will do the concatenation for us. That way it'll never be stale.

If `host` and `port` are immutable we should probably mark them as `const`.

The relationship between `host`, `subdomain` and `domain` raises the same question.

Reply:
Good point. `normalized` is not clear, I renamed to `host_port`.

All of these are string copies and end up getting freed. So even though they're not modified after the copy is created, they'd need to be casted to non-const when freeing.